File: algorithms/binary_search.py
def binary_search(arr, item):
	"""return index of item or place (-1) for insert """
	left = -1
	right = len(arr)
	while right - left > 1:
		middle = (right + left) // 2
		if item > arr[middle]:
			left = middle
		elif item < arr[middle]:
			right = middle
		else:
			return middle
	return left


def bin_search(arr, item):
	""" if item in arr that return index, else return -1"""
	left = -1
	right = len(arr)
	while right - left > 1:
		middle = (right + left) // 2
		if arr[middle] > item:
			right = middle
		elif arr[middle] < item:
			left = middle
		else:
			return middle
	return -1


def first_entry(arr, item):
	"""return index of first entry or -1"""
	left = 0
	right = len(arr) 
	while right - left > 1:
		middle  = (right + left) // 2
		if arr[middle] >= item:
			right = middle
		else:
			left = middle + 1
	if arr[right] == item:
		return right
	else:
		return -1

File: algorithms/binary_search.cpp
#include <iostream>

int main()
{
	// preparing data 
	int request, length_arr;
	std::cin >> request >> length_arr;
	int arr[length_arr];
	for (int i = 0; i < length_arr; ++i)
	{
		std::cin >> arr[i];
	}
	// binary search
	int middle, left = 0, right = length_arr;
	while (right - left > 1) {
		middle = (right + left) / 2;
		if (request > arr[middle]){
		  	left = middle + 1;
		} else if (request < arr[middle]){
			right = middle;
		} else {
			std::cout << middle;
			return 0; 
		}
	}
	std::cout << left;
	return 0;
}